REGRESSION(276827@main): A large animated image will flicker if it's displayed multiple times with different sizes

shallawa opened the pull request:
#### bd4296e26d45bccca513cdf21af3a1f8fa51c229
<pre>
REGRESSION(276827@main): A large animated image will flicker if it&apos;s displayed multiple times with different sizes
<a href="https://bugs.webkit.org/show_bug.cgi?id=274671">https://bugs.webkit.org/show_bug.cgi?id=274671</a>
<a href="https://rdar.apple.com/128557319">rdar://128557319</a>

Reviewed by Simon Fraser.

If the image is an animated image and the current frame is not available, we have
no choice but to decode it synchronously. Otherwise, a flicker will happen. This
was the BitmapImage::draw() behavior before 276827@main.

* LayoutTests/fast/images/animated-large-image-different-dest-size-expected.txt: Added.
* LayoutTests/fast/images/animated-large-image-different-dest-size.html: Added.
* LayoutTests/fast/images/resources/animated-red-green-blue-1000x1000-infinite.gif: Added.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::requestNativeImageAtIndexIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexRequestIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
(WebCore::BitmapImageSource::decodeCountForTesting const): Deleted.
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::blankDrawCountForTesting const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageBlankDrawCount):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/279483@main">https://commits.webkit.org/279483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00b9fe1ea8b5c14155966c4d1f02d0a3a6f4f1e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53586 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56866 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4316 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55892 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40418 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4120 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43437 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2828 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55684 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46315 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24572 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28001 "Found unexpected failure with change (failure)") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3636 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; Running layout-tests; Running re-run-layout-tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2468 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3808 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58463 "Built successfully") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3860 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50845 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46487 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11687 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29724 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/BitmapImageSource.cpp:
```diff
@@ -526,8 +526,10 @@ DecodingStatus BitmapImageSource::requestNativeImageAtIndexIfNeeded(unsigned ind
     if (index >= m_frames.size())
         return DecodingStatus::Invalid;
 
+    // Never decode the same frame from two different threads.
```
We need a radar to reference here.

shallawa replied:
I added a FIXME comment to remove this completely. ImageIO is supposed to handle this case safely and be thread safe when decoding the same frame from multiple threads.

shallawa committed:
REGRESSION(276827@main): A large animated image will flicker if it's displayed multiple times with different sizes
https://bugs.webkit.org/show_bug.cgi?id=274671
rdar://128557319

Reviewed by Simon Fraser.

If the image is an animated image and the current frame is not available, we have
no choice but to decode it synchronously. Otherwise, a flicker will happen. This
was the BitmapImage::draw() behavior before 276827@main.

* LayoutTests/fast/images/animated-large-image-different-dest-size-expected.txt: Added.
* LayoutTests/fast/images/animated-large-image-different-dest-size.html: Added.
* LayoutTests/fast/images/resources/animated-red-green-blue-1000x1000-infinite.gif: Added.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::requestNativeImageAtIndexIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexRequestIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
(WebCore::BitmapImageSource::decodeCountForTesting const): Deleted.
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::blankDrawCountForTesting const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageBlankDrawCount):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/279483@main
